### PR TITLE
Fix parsing of Between statements

### LINKED ISSUE #478: Can't parse complex 'between' statement
test.sql
```sql
select
    my_table.*
from
    my_table
    inner join my_other_table
        on my_table.key = my_other_table.key
where
    my_table.field_1 between my_other_table.field_2 - 10 and my_other_table.field_3
```

lint result
```sh
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   8 | P:  21 |  PRS | Found unparsable section: ' between
                       | my_other_table.field_2 - 10 and...'
WARNING: Parsing errors found and dialect is set to 'ansi'. Have you configured your dialect?
```

This also happens if the complex `between` statement is in a `join` statement. The issue goes away if simplify  `my_other_table.field_2 - 10 ` to just `my_other_table.field_2`

### PR DESCRIPTION
**Pull request by alanmcruickshank**
Fixes #478 .

Also covers two bugs found during debugging:
- This also improves pruning during the parsing set which should result in a performance improvement
- A bug in the logging routines which meant that the `match_depth` we unintentionally being zeroed.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 1]
```diff
@@ -939,9 +939,24 @@ class CaseExpressionSegment(BaseSegment):
                 Sequence(
                     Ref.keyword('NOT', optional=True),
                     'BETWEEN',
+                    # In a between expression, we're restricted to arithmetic operations
+                    # which means we can look for some binary operations without worrying
+                    # about the binary ones (and therefore AND).
```
Comment (pwildenhain) [inline src/sqlfluff/dialects/dialect_ansi.py:978]: > look for some binary operations without worrying about the binary ones

The use of "binary" here twice confuses me 🤔 could you clarify this (💯 love your detailed commenting style btw)

**Review by pwildenhain — CHANGES_REQUESTED** [review]
Thanks for getting to this so quickly! Is there any spot where we can add a test case for this?
Perhaps here: https://github.com/sqlfluff/sqlfluff/blob/ac/bugfix/test/fixtures/parser/ansi/select_o.sql

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_ansi.py:978]: I don't think what I've written actually makes sense! Sorry about that - clearly written while sleepy 😴 .

I shall reword it to be more useful.

**Comment by alanmcruickshank** [comment]
@pwildenhain - can you re-review based on the updated test case and the reworded comments?

**Review by pwildenhain — APPROVED** [review]
👏 🙌 Nicely done
